examples/text_analysis_mosaicml_mlflow.py

```diff
@@ -0,0 +1,161 @@
+# MAGIC 2. Run text analysis with MosaicML llama2-70b model & Databricks MLflow
+# MAGIC
+# MAGIC **NOTE**: This notebook requires >= DBR 13.3 LTS ML Runtime
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Install dependencies
+
+# COMMAND ----------
+
+# MAGIC %pip install mlflow[gateway]
+# MAGIC dbutils.library.restartPython()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Setup widgets
+
+# COMMAND ----------
+
+dbutils.widgets.text("from_tables", "discoverx_sample.*.*", "from tables")
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Import required libs and initialize variables
+
+# COMMAND ----------
+
+import pandas as pd
+from pyspark.sql.functions import (
+    pandas_udf,
+    col,
+    concat,
+    lit,
+    explode,
+    count,
+    avg,
+    min,
+    max,
+    sum,
+    collect_set,
+    concat_ws,
+)
+from pyspark.sql.types import ArrayType, StringType, StructType, FloatType, StructField
+from typing import Iterator
+
+# COMMAND ----------
+
+from_tables = dbutils.widgets.get("from_tables")
+
+# Set the sample rows size
+sample_size = 100
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Initialize discoverx
+
+# COMMAND ----------
+
+from discoverx import DX
+
+dx = DX()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Transform all sampled tables
+
+# COMMAND ----------
+
+unpivoted_df = (
+    dx.from_tables(from_tables)
+    .unpivot_string_columns(sample_size=sample_size)
+    .to_union_dataframe()
+    .localCheckpoint()  # Checkpointing to reduce the query plan size
+)
+
+# unpivoted_df.display()
+
+# COMMAND ----------
+
+display(unpivoted_df)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Define udf to use MosiacML model
+
+# COMMAND ----------
+
+# get or create mosaic route
+import mlflow
+from mlflow import gateway
+
+gateway.set_gateway_uri(gateway_uri="databricks")
+
+mosaic_route_name = "mosaicml-llama2-70b-completions"
+
+try:
+    route = gateway.get_route(mosaic_route_name)
+except:
+    # Create a route for embeddings with MosaicML
+    print(f"Creating the route {mosaic_route_name}")
+    print(
+        gateway.create_route(
+            name=mosaic_route_name,
+            route_type="llm/v1/completions",
+            model={
+                "name": "llama2-70b-chat",
+                "provider": "mosaicml",
+                "mosaicml_config": {"mosaicml_api_key": dbutils.secrets.get(scope="dbdemos", key="mosaic_ml_api_key")},
+            },
+        )
+    )
+
+# COMMAND ----------
+
+from mlflow import gateway
+
+@pandas_udf(StringType())
+def predict_value_udf(s):
+    def predict_value(s):
+        data = {
+            "prompt": f""" [INST] 
+        <<SYS>>
+        Reply with either YES or NO
+        <</SYS>>
+        Is this news article related to aquisition/merger ?
+        News Article: {s}
+         [/INST]
+        """
+        }
+        r = mlflow.gateway.query(route=mosaic_route_name, data=data)
```

Comment:
This can become quite expensive if we pass a lot of data correct? Would it make sense to also show how to estimate the cost before actually sending the queries? For this we would need to estimate the number of tokens which would probably again require sending queries to the api for embeddings. But maybe we can do at least a rough estimate using the number of words. What do you think?

Reply:
Very good point @david-tempelmann .
I think we could estimate the number of tokens by multiplying the number of characters by 4, it would not be perfect but it would give a good estimate.

Reply:
Added a link in notes for pricing . Not sure if it is good idea to calculate cost with some udf